### ISSUE TITLE
Backprop zeroing on FullyConnLayer

### ISSUE THREAD

**Issue by lakowske** [karpathy/convnetjs#62]
I'm curious why the input gradients are zero'd out in FullyConnLayer.backward() and not the biases and filter gradients.

So zero out input, filter and bias gradients like this:
```    
backward: function() {
      var V = this.in_act;
      V.dw = global.zeros(V.w.length); // zero out the gradient in input Vol                                                                                                                                                               

      // compute gradient wrt weights and data                                                                                                                                                                                             
      for(var i=0;i<this.out_depth;i++) {
          var tfi = this.filters[i];
          tfi.dw = global.zeros(tfi.dw.length); // zero out the gradient in filter Vol                                                                                                                                                      
          this.biases.dw[i] = 0;
          var chain_grad = this.out_act.dw[i];
          for(var d=0;d<this.num_inputs;d++) {
              V.dw[d] += tfi.w[d]*chain_grad; // grad wrt input data                                                                                                                                                                       
              tfi.dw[d] += V.w[d]*chain_grad; // grad wrt params                                                                                                                                                                           
          }
          this.biases.dw[i] += chain_grad;
      }
```

Instead of the original:
```
    backward: function() {
      var V = this.in_act;
      V.dw = global.zeros(V.w.length); // zero out the gradient in input Vol
      
      // compute gradient wrt weights and data
      for(var i=0;i<this.out_depth;i++) {
        var tfi = this.filters[i];
        var chain_grad = this.out_act.dw[i];
        for(var d=0;d<this.num_inputs;d++) {
          V.dw[d] += tfi.w[d]*chain_grad; // grad wrt input data
          tfi.dw[d] += V.w[d]*chain_grad; // grad wrt params
        }
        this.biases.dw[i] += chain_grad;
      }
    }
```
When I use the original code, my 1 layer, 1 neuron net oscillates around the training set rather than steadily decreasing the cost.

When I zero the input, filter and gradient biases, I get steadily decreasing cost.  Isn't the filter gradient supposed to be wrt the input rather than the sum of past backprop gradients?

Thanks for helping me understand!
Seth

**Comment by andy-gh** [comment]
I think he is summing up the gradients over all training results for a given minibatch. The network is forward/back-propped against a number of samples from a minibatch, the gradients are summed up at each backward() pass, then averaged before weight update. The averaging happens here:
convnetjs_trainers.js, line 91:
`gij = (l2grad + l1grad + g[j]) / this.batch_size; // raw batch gradient`

here's more on [minibatch grad update](http://stats.stackexchange.com/questions/183840/sum-or-average-of-gradients-in-mini-batch-gradient-decent).
